File: mamda/dotnet/src/examples/MamdaComboTicker/MamdaComboTicker.cs
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2011 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

using System;
using System.Collections;
using System.Threading;
using System.IO;

namespace Wombat.Mamda.Examples
{
	/// <summary>
	/// </summary>
	public class MamdaComboTicker
	{
        private static MamdaSubscription[] mamdaSubscriptions;
		private static int				   myQuietModeLevel;

		public static void Main(string[] args)
		{
			MamaTransport        transport    = null;
			MamaQueue            defaultQueue = null;
			MamaDictionary       dictionary   = null;
			CommandLineProcessor options      = new CommandLineProcessor(args);
			myQuietModeLevel = options.getQuietModeLevel();

			if (options.hasLogLevel())
			{
				myLogLevel = options.getLogLevel();
				Mama.enableLogging(myLogLevel);
			}

			try
			{
				// Initialize MAMDA
				myBridge = new MamaBridge(options.getMiddleware());
				Mama.open();
				transport = new MamaTransport();
				transport.create(options.getTransport(), myBridge);
				defaultQueue = Mama.getDefaultEventQueue(myBridge);

				/*Get the data dictionary*/
				MamaSource dictionarySource = new MamaSource();
                dictionarySource.symbolNamespace = options.getDictSource();
				dictionarySource.transport = transport;
				dictionary = buildDataDictionary(transport, defaultQueue, dictionarySource);

				MamdaQuoteFields.setDictionary(dictionary, null);
				MamdaTradeFields.setDictionary(dictionary, null);
                MamdaSecurityStatusFields.setDictionary(dictionary, null);

                mamdaSubscriptions = new MamdaSubscription [options.getSymbolList().Count];
                int i=0;
				foreach (string symbol in options.getSymbolList())
				{
					mamdaSubscriptions[i] =  new MamdaSubscription();

					MamdaTradeListener aTradeListener = new MamdaTradeListener();
					MamdaQuoteListener aQuoteListener = new MamdaQuoteListener();
                    MamdaSecurityStatusListener aSecurityStatusListener = new MamdaSecurityStatusListener();
					ComboTicker        aTicker        = new ComboTicker();

					aTradeListener.addHandler(aTicker);
					aQuoteListener.addHandler(aTicker);
                    aSecurityStatusListener.addHandler(aTicker);
					mamdaSubscriptions[i].addMsgListener(aTradeListener);
					mamdaSubscriptions[i].addMsgListener(aQuoteListener);
					mamdaSubscriptions[i].addStaleListener(aTicker);
					mamdaSubscriptions[i].addErrorListener(aTicker);

					mamdaSubscriptions[i].create(transport,
						defaultQueue,
						options.getSource(),
						symbol,
						null/*Closure*/);

                    i++;
				}

				Console.WriteLine("Hit Enter or Ctrl-C to exit.");
				Mama.start(myBridge);
				GC.KeepAlive(dictionary);
				Console.ReadLine();
			}
			catch (Exception e)
			{
				Console.Error.WriteLine(e.ToString());
				Environment.Exit(1);
			}
		}

		private class ComboTicker : MamdaTradeHandler,
									MamdaQuoteHandler,
									MamdaSecurityStatusHandler,
									MamdaStaleListener,
									MamdaErrorListener
		{
			public void onTradeRecap(
				MamdaSubscription   sub,
				MamdaTradeListener  listener,
				MamaMsg             msg,
				MamdaTradeRecap     recap)
			{
				prettyPrint("Trade Recap ({0}): ", sub.getSymbol());
			}

			public void onTradeReport(
				MamdaSubscription   sub,
				MamdaTradeListener  listener,
				MamaMsg             msg,
				MamdaTradeReport    trade,
				MamdaTradeRecap     recap)
			{
				prettyPrint(
					"Trade ({0}:{1}): {2} @ {3} (seq#: {4}; time: {5}; acttime: {6}; qual: {7})",
					sub.getSymbol(),
					recap.getTradeCount(),
					trade.getTradeVolume(),
					trade.getTradePrice(),
					trade.getEventSeqNum(),
					trade.getEventTime(),
					trade.getActivityTime(),
					trade.getTradeQual());
			}

			public void onTradeGap (
				MamdaSubscription   sub,
				MamdaTradeListener  listener,
				MamaMsg             msg,
				MamdaTradeGap       gap,
				MamdaTradeRecap     recap)
			{
				prettyPrint("Trade gap ({0}-{1})",
					gap.getBeginGapSeqNum(),
					gap.getEndGapSeqNum());
			}

			public void onTradeCancelOrError (
				MamdaSubscription        sub,
				MamdaTradeListener       listener,
				MamaMsg                  msg,
				MamdaTradeCancelOrError  cancelOrError,
				MamdaTradeRecap          recap)
			{
				prettyPrint("Trade error/cancel ({0}):", sub.getSymbol());
			}

			public void onTradeCorrection (
				MamdaSubscription        sub,
				MamdaTradeListener       listener,
				MamaMsg                  msg,
				MamdaTradeCorrection     correction,
				MamdaTradeRecap          recap)
			{
				prettyPrint("Trade correction ({0}):", sub.getSymbol());
			}

			public void onTradeClosing (
				MamdaSubscription        sub,
				MamdaTradeListener       listener,
				MamaMsg                  msg,
				MamdaTradeClosing        closing,
				MamdaTradeRecap          recap)
			{
				prettyPrint("Trade closing ({0}):", sub.getSymbol());
			}

			public void onQuoteRecap(
				MamdaSubscription   subscription,
				MamdaQuoteListener  listener,
				MamaMsg             msg,
				MamdaQuoteRecap     recap)
			{
				prettyPrint("Quote Recap ({0})", subscription.getSymbol());
			}

			public void onQuoteUpdate(
				MamdaSubscription   subscription,
				MamdaQuoteListener  listener,
				MamaMsg             msg,
				MamdaQuoteUpdate    update,
				MamdaQuoteRecap     recap)
			{
				string format = "Quote ({0}:{1}): {2} {3} X {4} {5} (seq#: {6}; time: {7}; qual: {8})";
				prettyPrint(format,
					subscription.getSymbol(),
					recap.getQuoteCount(),
					update.getBidPrice(),
					update.getBidSize(),
					update.getAskSize(),
					update.getAskPrice(),
					update.getEventSeqNum(),
					update.getEventTime(),
					update.getQuoteQual());
			}

			public void onQuoteGap(
				MamdaSubscription   subscription,
				MamdaQuoteListener  listener,
				MamaMsg             msg,
				MamdaQuoteGap       gap,
				MamdaQuoteRecap     recap)
			{
				prettyPrint("Quote gap ({0}-{1})",
					gap.getBeginGapSeqNum(),
					gap.getEndGapSeqNum());
			}

			public void onQuoteClosing(
				MamdaSubscription   subscription,
				MamdaQuoteListener  listener,
				MamaMsg             msg,
				MamdaQuoteClosing   quoteClosing,
				MamdaQuoteRecap     recap)
			{
				prettyPrint("Quote closing ({0})", subscription.getSymbol());
			}

			public void onSecurityStatusRecap(
				MamdaSubscription            subscription,
				MamdaSecurityStatusListener  listener,
				MamaMsg                      msg,
				MamdaSecurityStatusRecap     recap)
			{
				prettyPrint("SecStatus recap ({0} - {1})", subscription.getSymbol(), recap.getSecurityStatusStr ());
 			}

			public void onSecurityStatusUpdate(
				MamdaSubscription            subscription,
				MamdaSecurityStatusListener  listener,
				MamaMsg                      msg,
				MamdaSecurityStatusUpdate    update,
				MamdaSecurityStatusRecap     recap)
			{
				prettyPrint("SecStatus update ({0} - {1}", subscription.getSymbol (), listener.getSecurityStatusStr ());
			}

			public void onStale(
				MamdaSubscription   subscription,
				mamaQuality         quality)
			{
				prettyPrint("Stale ({0} - {1})", subscription.getSymbol(), quality);
			}

			public void onError(
				MamdaSubscription   subscription,
				MamdaErrorSeverity  severity,
				MamdaErrorCode      errorCode,
				string              errorStr)
			{
				prettyPrint("Error ({0}): {1} ({2})",
					subscription.getSymbol(),
					MamdaErrorCodes.stringForMamdaError(errorCode),
					errorStr);
			}

			private static void prettyPrint(string format, params object[] args)
			{
				if (myQuietModeLevel<1)
				{
					Console.WriteLine(format, args);
				}
			}
		}

		private static MamaDictionary buildDataDictionary(
			MamaTransport transport,
			MamaQueue defaultQueue,
			MamaSource dictionarySource)
		{
			bool[] gotDict = new bool[] { false };
			MamaDictionaryCallback dictionaryCallback = new DictionaryCallback(gotDict);
			lock (dictionaryCallback)
			{
				MamaSubscription subscription = new MamaSubscription ();

				MamaDictionary dictionary = new MamaDictionary();
				dictionary.create(
					defaultQueue,
					dictionaryCallback,
					dictionarySource,
					3,
					10);

				Mama.start(myBridge);
				if (!gotDict[0])
				{
					if (!Monitor.TryEnter(dictionaryCallback, 30000))
					{
						Console.Error.WriteLine("Timed out waiting for dictionary.");
						Environment.Exit(0);
					}
					Monitor.Exit(dictionaryCallback);
				}
				return dictionary;
			}
		}

		private class DictionaryCallback : MamaDictionaryCallback
		{
			public DictionaryCallback(bool[] gotDict)
			{
				gotDict_ = gotDict;
			}

			public void onTimeout(MamaDictionary dictionary)
			{
				Console.Error.WriteLine("Timed out waiting for dictionary");
				Environment.Exit(1);
			}

			public void onError(MamaDictionary dictionary, string message)
			{
				Console.Error.WriteLine("Error getting dictionary: {0}", message);
				Environment.Exit(1);
			}

			public void onComplete(MamaDictionary dictionary)
			{
				lock (this)
				{
					gotDict_[0] = true;
					Mama.stop(myBridge);
					Monitor.PulseAll(this);
				}
			}

			private bool[] gotDict_;
		}

		private static MamaLogLevel myLogLevel = MamaLogLevel.MAMA_LOG_LEVEL_NORMAL;
		private static MamaBridge	myBridge   = null;
	}
}

File: mamda/java/com/wombat/mamda/examples/MamdaComboTicker.java
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2012 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

package com.wombat.mamda.examples;

import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.Iterator;
import com.wombat.mamda.*;
import com.wombat.mama.Mama;
import com.wombat.mama.MamaBridge;
import com.wombat.mama.MamaMsg;
import com.wombat.mama.MamaTransport;
import com.wombat.mama.MamaDictionary;
import com.wombat.mama.MamaDictionaryCallback;
import com.wombat.mama.MamaSubscription;
import com.wombat.mama.MamaSource;

public class MamdaComboTicker
{
    private static Logger theLogger = 
                              Logger.getLogger( "com.wombat.mamda.examples" );

    private static MamaSource   mMamaSource ;
    private static MamaBridge   mBridge = null;
    private static Level        mLogLevel = Level.INFO;
    private static int          mQuietModeLevel = 0;
    private static String       mDictTransportName = null;

    private static MamaDictionary buildDataDictionary (MamaTransport transport,
                                                       String        dictSource)
        throws InterruptedException
    {
        final boolean gotDict[] = {false};
        MamaDictionaryCallback dictionaryCallback =
            new MamaDictionaryCallback ()
        {
            public void onTimeout ()
            {
                System.err.println
                    ("Timed out waiting for dictionary");
                System.exit (1);
            }

            public void onError (final String s)
            {
                System.err.println ("Error getting dictionary: " + s);
                System.exit (1);
            }

            public synchronized void onComplete ()
            {
                gotDict[0] = true;
                Mama.stop (mBridge);
                notifyAll ();
            }
        };

        synchronized (dictionaryCallback)
        {
            MamaSubscription subscription = new MamaSubscription ();
            
            mMamaSource.setTransport       (transport);
            mMamaSource.setSymbolNamespace (dictSource);
            MamaDictionary   dictionary   = subscription.
                createDictionarySubscription (
                    dictionaryCallback,
                    Mama.getDefaultQueue (mBridge),
                    mMamaSource);

            Mama.start(mBridge);
            if (!gotDict[0]) dictionaryCallback.wait (30000);
            if (!gotDict[0])
            {
                System.err.println ("Timed out waiting for dictionary.");
                System.exit (0);
            }
            return dictionary;
        }
    }

    public static void main (final String[] args)
    {
        MamaTransport        transport      = null;
	    MamaTransport	     mDictTransport = null;
        MamaDictionary       dictionary     = null;
        CommandLineProcessor options        = new CommandLineProcessor (args);

        mQuietModeLevel = options.getQuietModeLevel ();
        mLogLevel       = options.getLogLevel ();
        if (mLogLevel != null)
        {
            theLogger.setLevel (mLogLevel);
            Mama.enableLogging (mLogLevel);
        }
        
        theLogger.info ("Source: " + options.getSource());

        try
        {
            // Initialize MAMA
            mBridge = options.getBridge();
            Mama.open ();
            transport = new MamaTransport ();
            transport.create (options.getTransport(), mBridge);
            mMamaSource = new MamaSource ();

            mDictTransportName = options.getDictTransport();

            if (mDictTransportName != null)
            {
                mDictTransport = new MamaTransport ();
                mDictTransport.create (mDictTransportName, mBridge);
            }
            else 
            {
                mDictTransport = transport;
            }

            //Get the Data dictionary.....
            dictionary = buildDataDictionary (mDictTransport,options.getDictSource());
            
            MamdaQuoteFields.setDictionary          (dictionary, null);
            MamdaTradeFields.setDictionary          (dictionary, null);
            MamdaCommonFields.setDictionary         (dictionary, null);
            MamdaSecurityStatusFields.setDictionary (dictionary, null);

            for (Iterator iterator = options.getSymbolList().iterator();
                 iterator.hasNext();)
            {
                final String symbol = (String) iterator.next();
                MamdaSubscription  aSubscription =  new MamdaSubscription ();
                
                MamdaTradeListener aTradeListener = new MamdaTradeListener ();
                MamdaQuoteListener aQuoteListener = new MamdaQuoteListener ();
                ComboTicker        aTicker        = new ComboTicker ();
                MamdaSecurityStatusListener aSecurityStatusListener = 
                    new MamdaSecurityStatusListener();

                aTradeListener.addHandler          (aTicker);
                aQuoteListener.addHandler          (aTicker);
                aSecurityStatusListener.addHandler (aTicker);
                aSubscription.addMsgListener       (aTradeListener);
                aSubscription.addMsgListener       (aQuoteListener);
                aSubscription.addStaleListener     (aTicker);
                aSubscription.addErrorListener     (aTicker);

                aSubscription.create (transport,
                                      Mama.getDefaultQueue (mBridge),
                                      options.getSource (),
                                      symbol,
                                      null /* Closure */);

                theLogger.fine ("Subscribed to: " + symbol);
            }

            Mama.start (mBridge);
            Thread.currentThread ().join ();
        }
        catch (Exception e)
        {
            e.printStackTrace ();
            System.exit (1);
        }
    }

    private static class ComboTicker implements MamdaTradeHandler, 
                                                MamdaQuoteHandler,
                                                MamdaSecurityStatusHandler,
                                                MamdaStaleListener,
                                                MamdaErrorListener 
   {
        public void onTradeRecap (
            final MamdaSubscription   sub,
            final MamdaTradeListener  listener,
            final MamaMsg             msg,
            final MamdaTradeRecap     recap)
        {
            prettyPrint ("Trade Recap (" + sub.getSymbol () + ")");
        }

        public void onTradeReport (
            final MamdaSubscription   sub,
            final MamdaTradeListener  listener,
            final MamaMsg             msg,
            final MamdaTradeReport    trade,
            final MamdaTradeRecap     recap)
        {
            prettyPrint ("Trade ("  + sub.getSymbol () +
                         ":"        + recap.getTradeCount ()    +
                         "):  "     + trade.getTradeVolume ()   +
                         " @ "      + trade.getTradePrice ()    +
                         " (seq#: " + trade.getEventSeqNum ()   +
                         "; time: " + trade.getEventTime ()     +
                         "; qual: " + trade.getTradeQual ()     + 
                         "; acttime: " + trade.getActivityTime () +
                         "; tradeCount:"+ recap.getTradeCount () + ")");
        }

        public void onTradeGap (
            final MamdaSubscription   sub,
            final MamdaTradeListener  listener,
            final MamaMsg             msg,
            final MamdaTradeGap       event,
            final MamdaTradeRecap     recap)
        {
            prettyPrint ("Trade gap (" + event.getBeginGapSeqNum () +
                         "-" + event.getEndGapSeqNum () + ")");
        }

        public void onTradeCancelOrError (
            final MamdaSubscription        sub,
            final MamdaTradeListener       listener,
            final MamaMsg                  msg,
            final MamdaTradeCancelOrError  event,
            final MamdaTradeRecap          recap)
        {
            prettyPrint ("Trade error/cancel (" + sub.getSymbol() + ")");
        }

        public void onTradeCorrection (
            final MamdaSubscription        sub,
            final MamdaTradeListener       listener,
            final MamaMsg                  msg,
            final MamdaTradeCorrection     event,
            final MamdaTradeRecap          recap)
        {
            prettyPrint ("Trade correction (" + sub.getSymbol() + ")");
        }

        public void onTradeClosing (
            final MamdaSubscription        sub,
            final MamdaTradeListener       listener,
            final MamaMsg                  msg,
            final MamdaTradeClosing        event,
            final MamdaTradeRecap          recap)
        {
            prettyPrint ("Trade closing (" + sub.getSymbol() + ")");
        }

        public void onQuoteRecap (
            final MamdaSubscription   sub,
            final MamdaQuoteListener  listener,
            final MamaMsg             msg,
            final MamdaQuoteRecap     recap)
        {
            prettyPrint ("Quote Recap (" + sub.getSymbol() + "): ");
        }

        public void onQuoteUpdate (
            final MamdaSubscription   sub,
            final MamdaQuoteListener  listener,
            final MamaMsg             msg,
            final MamdaQuoteUpdate    event,
            final MamdaQuoteRecap     recap)
        {
            prettyPrint ("Quote ("  + sub.getSymbol()        +
                         ":"        + recap.getQuoteCount()  +
                         "):  "     + event.getBidPrice()    +
                         " "        + event.getBidSize()     +
                         " X "      + event.getAskSize()     +
                         " "        + event.getAskPrice()    +
                         " (seq#: " + event.getEventSeqNum() +
                         "; time: " + event.getEventTime()   +
                         "; qual: " + event.getQuoteQual()   +
                         "; quote count: " + recap.getQuoteCount ()+ ")");
        }

        public void onQuoteGap (
            final MamdaSubscription   sub,
            final MamdaQuoteListener  listener,
            final MamaMsg             msg,
            final MamdaQuoteGap       event,
            final MamdaQuoteRecap     recap)
        {
            prettyPrint ("Quote gap (" + event.getBeginGapSeqNum() +
                         "-" + event.getEndGapSeqNum() + ")");
        }

        public void onQuoteClosing (
            final MamdaSubscription   sub,
            final MamdaQuoteListener  listener,
            final MamaMsg             msg,
            final MamdaQuoteClosing   event,
            final MamdaQuoteRecap     recap)
        {
            prettyPrint ("Quote gap (" + sub.getSymbol () + ")");
        }

        public void onSecurityStatusRecap (
            MamdaSubscription            subscription,
            MamdaSecurityStatusListener  listener,
            MamaMsg                      msg,
            MamdaSecurityStatusRecap     recap)
        {
            prettyPrint ("SecStatus recap (" + subscription.getSymbol () + 
                         " Status:" + recap.getSecurityStatusStr () + recap.getSecurityStatusQualifierStr ()+ ")" );

        }

        public void onSecurityStatusUpdate (
            MamdaSubscription            subscription,
            MamdaSecurityStatusListener  listener,
            MamaMsg                      msg,
            MamdaSecurityStatusUpdate    event,
            MamdaSecurityStatusRecap     recap)
        {
            System.out.println ("SecStatus update (" + subscription.getSymbol () +
                                " Status:" + recap.getSecurityStatusStr () + recap.getSecurityStatusQualifierStr ()+ ")");
        }


        public void onStale (
            MamdaSubscription   subscription,
            short               quality)
        {
            prettyPrint ("Stale (" + subscription.getSymbol () + ")");
        }

        public void onError (
            MamdaSubscription   subscription,
            short               severity,
            short               errorCode,
            String              errorStr)
        {
            prettyPrint ("Error (" + subscription.getSymbol () + ")");
        }

        public void prettyPrint (String what)
        {
            if (mQuietModeLevel<1)
            {
                System.out.println(what);
            }
        }
    }
}
